src/libraries/System.Private.CoreLib/src/System/MemoryExtensions.Split.cs

```diff
@@ -0,0 +1,152 @@
+        /// <returns>Returns a <see cref="System.MemoryExtensions.SpanSplitEnumerator{T}"/>.</returns>
+        public static SpanSplitEnumerator<T> Split<T>(this ReadOnlySpan<T> source, ReadOnlySpan<T> separator)
+            where T : IEquatable<T> => new SpanSplitEnumerator<T>(source, separator, treatAsSingleSeparator: true);
+
+        /// <summary>
+        /// Returns a type that allows for enumeration of each element within a split span
+        /// using any of the provided elements.
+        /// </summary>
+        /// <param name="source">The source span to be enumerated.</param>
+        /// <param name="separators">The separators to be used to split the provided span.</param>
+        /// <returns>Returns a <see cref="System.MemoryExtensions.SpanSplitEnumerator{T}"/>.</returns>
+        public static SpanSplitEnumerator<T> SplitAny<T>(this ReadOnlySpan<T> source, [UnscopedRef] params ReadOnlySpan<T> separators)
+            where T : IEquatable<T> => new SpanSplitEnumerator<T>(source, separators, treatAsSingleSeparator: false);
+
+        /// <summary>
+        /// Returns a type that allows for enumeration of each element within a split span
+        /// using the provided <see cref="System.MemoryExtensions.SpanSplitEnumerator{T}"/>.
+        /// </summary>
+        /// <param name="source">The source span to be enumerated.</param>
+        /// <param name="separators">The <see cref="System.MemoryExtensions.SpanSplitEnumerator{T}"/> to be used to split the provided span.</param>
+        /// <returns>Returns a <see cref="System.MemoryExtensions.SpanSplitEnumerator{T}"/>.</returns>
+        public static SpanSplitEnumerator<T> SplitAny<T>(this ReadOnlySpan<T> source, SearchValues<T> separators)
+            where T : IEquatable<T> => new SpanSplitEnumerator<T>(source, separators);
+
+        /// <summary>
+        /// <see cref="System.MemoryExtensions.SpanSplitEnumerator{T}"/> allows for enumeration of each element within a <see cref="System.ReadOnlySpan{T}"/>
+        /// that has been split using a provided separator.
+        /// </summary>
+        public ref struct SpanSplitEnumerator<T> where T : IEquatable<T>
+        {
+            private enum SplitMode
+            {
+                None = 0,
+                SingleToken,
+                Sequence,
+                Any,
+                SearchValues
+            }
+
+            private readonly ReadOnlySpan<T> _span;
+
+            private readonly T _separator = default!;
+            private readonly ReadOnlySpan<T> _separatorBuffer;
+            private readonly SearchValues<T> _searchValues = default!;
+
+            private readonly int _separatorLength;
+            private readonly SplitMode _splitMode;
+
+            private readonly bool _isInitialized = true;
+
+            private int _startCurrent = 0;
+            private int _endCurrent = 0;
+            private int _startNext = 0;
```

Comment:
Is _startNext needed rather than just updating _span to be the remainder to be searched?

Reply:
As with https://github.com/dotnet/runtime/pull/104534/files#r1668846871 I believe this was PR feedback at the time. Happy to change it

Reply:
Had a look, it's slightly tricky to just use the remainder of the span to continue searching. There needs to be something that allows you to distinguish between the first `MoveNext()` call and the `MoveNext()` call that terminates the search. Having a simple `_span.Length == 0` exit condition would be hit on the first `MoveNext()` when the source `_span` ::= `{ Length: 0 }`. It wouldn't be impossible to rewrite it to use the remainder instead, but you would need to store _some_ state that `MoveNext()` has at least once been called.